pkg/storage/stores/shipper/indexshipper/gatewayclient/gateway_client.go

```diff
@@ -122,47 +131,113 @@ func NewGatewayClient(cfg IndexGatewayClientConfig, r prometheus.Registerer, lim
 		ring:                              cfg.Ring,
 		limits:                            limits,
+		done:                              make(chan struct{}),
 	}
 
 	dialOpts, err := cfg.GRPCClientConfig.DialOption(grpcclient.Instrument(sgClient.storeGatewayClientRequestDuration))
 	if err != nil {
 		return nil, errors.Wrap(err, "index gateway grpc dial option")
 	}
+	factory := func(addr string) (ring_client.PoolClient, error) {
+		igPool, err := NewIndexGatewayGRPCPool(addr, dialOpts)
+		if err != nil {
+			return nil, errors.Wrap(err, "new index gateway grpc pool")
+		}
+
+		return igPool, nil
+	}
+
+	//FIXME(ewelch) we don't expose the pool configs nor set defaults, and register flags is kind of messed up with remote config being defined somewhere else
+	//make a separate PR to make the pool config generic so it can be used with proper names in multiple places.
+	sgClient.cfg.PoolConfig.RemoteTimeout = 2 * time.Second
+	sgClient.cfg.PoolConfig.ClientCleanupPeriod = 5 * time.Second
+	sgClient.cfg.PoolConfig.HealthCheckIngesters = true
 
 	if sgClient.cfg.Mode == indexgateway.RingMode {
-		factory := func(addr string) (ring_client.PoolClient, error) {
-			igPool, err := NewIndexGatewayGRPCPool(addr, dialOpts)
+		sgClient.pool = clientpool.NewPool(sgClient.cfg.PoolConfig, sgClient.ring, factory, logger)
+	} else {
+		// Note we don't use clientpool.NewPool because we want to provide our own discovery function
+		poolCfg := ring_client.PoolConfig{
+			CheckInterval:      sgClient.cfg.PoolConfig.ClientCleanupPeriod,
+			HealthCheckEnabled: sgClient.cfg.PoolConfig.HealthCheckIngesters,
+			HealthCheckTimeout: sgClient.cfg.PoolConfig.RemoteTimeout,
+		}
+		clients := prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "loki",
+			Name:      "index_gateway_clients",
+			Help:      "The current number of index gateway clients.",
+		})
+		if r != nil {
+			err := r.Register(clients)
 			if err != nil {
-				return nil, errors.Wrap(err, "new index gateway grpc pool")
+				alreadyErr, ok := err.(prometheus.AlreadyRegisteredError)
+				if !ok {
+					return nil, err
+				}
+				clients = alreadyErr.ExistingCollector.(prometheus.Gauge)
 			}
-
-			return igPool, nil
 		}
+		//TODO(ewelch) we can't use metrics in the provider because of duplicate registration errors
+		dnsProvider := dns.NewProvider(logger, nil, dns.GolangResolverType)
+		sgClient.dnsProvider = dnsProvider
 
-		sgClient.pool = clientpool.NewPool(cfg.PoolConfig, sgClient.ring, factory, logger)
-	} else {
-		sgClient.conn, err = grpc.Dial(cfg.Address, dialOpts...)
-		if err != nil {
-			return nil, errors.Wrap(err, "index gateway grpc dial")
+		// Make an attempt to do one DNS lookup so we can start with addresses
+		sgClient.runDiscovery()
+		go sgClient.discoveryLoop()
+
+		discovery := func() ([]string, error) {
+			return dnsProvider.Addresses(), err
 		}
+		sgClient.pool = ring_client.NewPool("index gateway", poolCfg, discovery, factory, clients, logger)
```

Comment:
IMO configuring DNS discovery for the pool here belongs to a different place. Loki's wrapper around `ring_client.Pool` should be able to handle different discovery modes and metric names.